src/tools/rbd_wnbd/rbd_wnbd.cc

```diff
@@ -326,192 +330,53 @@ int send_map_request(std::string arguments) {
-    err = GetLastError();
-    derr << "CreateEvent failed: " << win32_strerror(err) << dendl;
-    exit_code = -ECHILD;
-    goto finally;
-  }
-  connect_o.hEvent = connect_event;
-  read_o.hEvent = read_event;
-
-  status = ConnectNamedPipe(pipe_handle, &connect_o);
-  err = GetLastError();
-  if (status || err != ERROR_IO_PENDING) {
-    if (status)
-      err = status;
-    derr << "ConnectNamedPipe failed: " << win32_strerror(err) << dendl;
-    exit_code = -ECHILD;
-    goto finally;
-  }
-  err = 0;
+  int argc;
+  std::vector<const char*> args;
 
-  dout(5) << __func__ << ": command arguments: " << arguments << dendl;
+  dout(5) << "Creating mapping using the same process. Command line: "
+          << command_line << dendl;
 
-  // We'll avoid running arbitrary commands, instead using the executable
-  // path of this process (expected to be the full rbd-wnbd.exe path).
-  err = get_exe_path(exe_path);
-  if (err) {
-    exit_code = -EINVAL;
-    goto finally;
-  }
-  command_line << std::quoted(exe_path)
-               << " " << std::regex_replace(arguments, pipe_pattern, "")
-               << " --pipe-name " << pipe_name.str();
-
-  dout(5) << __func__ << ": command line: " << command_line.str() << dendl;
-
-  GetStartupInfoW(&si);
-  // Create a detached child
-  if (!CreateProcessW(
-      NULL, const_cast<wchar_t*>(to_wstring(command_line.str()).c_str()),
-      NULL, NULL, FALSE, DETACHED_PROCESS,
-      NULL, NULL, &si, &pi)) {
-    err = GetLastError();
-    derr << "CreateProcess failed: " << win32_strerror(err) << dendl;
-    exit_code = -ECHILD;
-    goto finally;
+  // CommandLineToArgvW only has an UTF-16 variant.
+  LPWSTR* argv_w = CommandLineToArgvW(
+    to_wstring(command_line).c_str(), &argc);
+  if (!argv_w) {
+    DWORD err = GetLastError();
+    derr << "Couldn't parse args, error: "
+         << cpp_strerror(err) << dendl;
+    return -EINVAL;
   }
 
-  wait_events[0] = connect_event;
-  wait_events[1] = pi.hProcess;
-  status = WaitForMultipleObjects(2, wait_events, FALSE, timeout_ms);
-  switch(status) {
-    case WAIT_OBJECT_0:
-      if (!GetOverlappedResult(pipe_handle, &connect_o, &bytes_read, TRUE)) {
-        err = GetLastError();
-        derr << "Couldn't establish a connection with the child process. "
-             << "Error: " << win32_strerror(err) << dendl;
-        exit_code = -ECHILD;
-        goto clean_process;
-      }
-      // We have an incoming connection.
-      break;
-    case WAIT_OBJECT_0 + 1:
-      // The process has exited prematurely.
-      goto clean_process;
-    case WAIT_TIMEOUT:
-      derr << "Timed out waiting for child process connection." << dendl;
-      goto clean_process;
-    default:
-      derr << "Failed waiting for child process. Status: " << status << dendl;
-      goto clean_process;
+  std::vector<std::string> argv_sv;
+  argv_sv.reserve(argc);
+  for (int i=0; i<argc; i++) {
+    argv_sv.push_back(to_string(argv_w[i]));
```

Comment:
``args`` is a ``std::vector<const char*>``, having an additional ``std::vector<std::string>`` simplifies the  deallocations.

Reply:
Why is `reserve()` called only on `argv_sv`?  The relationship would be more obvious if these vectors were defined together and identical operations were performed on them:
```
  std::vector<std::string> argv_sv;
  std::vector<const char*> args;
  argv_sv.reserve(argc);
  args.reserve(argc);
  for (int i = 0; i < argc; i++) {
    argv_sv.push_back(to_string(argv_w[i]));
    args.push_back(argv_sv[i].c_str());
  }
```

Reply:
Makes sense, I can move the definitions. The reason for the ``reserve()`` call on ``argv_sv`` is that otherwise it gets resized as we push items and the ``argv_sv[i].c_str()`` pointers used by ``args`` then become invalid.

``args.reserve(argc);`` won't hurt, I can add that as well.

Reply:
> The reason for the `reserve()` call on `argv_sv` is that otherwise it gets resized as we push items and the `argv_sv[i].c_str()` pointers used by `args` then become invalid.

This sort of thing always deserves a comment.